tests-utils/src/main/scala/tech/beshu/ror/utils/misc/EsStartupChecker.scala

```diff
@@ -16,82 +16,85 @@
-        identity
-      )
-    }
+    retryBackoff(clusterIsReady(client), maxRetries = 30, interval = 2 seconds)
+      .map((_: Unit) => true)
+      .onErrorRecover(_ => false)
+      .runSyncUnsafe(2 minutes)
   }
 
-  private def retry(checkClusterHealthAction: => Boolean)
-                   (implicit startupThreshold: FiniteDuration) = {
-    val policy: RetryPolicy[Id] = limitRetriesByCumulativeDelay(startupThreshold, constantDelay(2 seconds))
-    val predicate = (_: Boolean) == true
-
-    @nowarn("cat=unused")
-    def onFailure(failedValue: Boolean, details: RetryDetails): Unit = {
-      logger.debug(s"[$name] ES not ready yet. Retrying ...")
-    }
-
-    retrying(policy, predicate, onFailure) {
-      checkClusterHealthAction
+  private def retryBackoff[A](source: Task[A],
+                              maxRetries: Int,
+                              interval: FiniteDuration): Task[A] = {
+    source.onErrorHandleWith {
+      case ex: Exception =>
+        if (maxRetries > 0)
+          retryBackoff(source, maxRetries - 1, interval).delayExecution(interval)
+        else
+          Task.raiseError(ex)
     }
   }
 
-  private def checkClusterHealth(client: RestClient) = {
-    val clusterHealthRequest = new HttpGet(client.from("_cluster/health"))
-    Try(client.execute(clusterHealthRequest)).bracket { response =>
-      response.getStatusLine.getStatusCode match {
-        case 200 =>
-          mode match {
-            case Mode.GreenCluster =>
-              val healthJson = deserializeJsonBody(RestClient.bodyFrom(response))
-              "green" == healthJson.get("status")
-            case Mode.Accessible =>
-              true
-          }
-        case _ =>
-          false
+  private def clusterIsReady(client: RestClient): Task[Unit] = {
+    Task
+      .delay(client.execute(new HttpGet(client.from("_cluster/health"))))
+      .recoverWith { ex =>
+        logger.error(s"[$name] ES not ready yet, healthcheck failed")
+        Task.raiseError(ex)
+      }
+      .flatMap { response =>
+        val result = response.getStatusLine.getStatusCode match {
+          case 200 =>
+            mode match {
+              case Mode.GreenCluster =>
+                val healthJson = deserializeJsonBody(RestClient.bodyFrom(response))
+                val healthStatus = healthJson.get("status")
+                if (healthStatus == "green") {
+                  logger.info(s"[$name] ES is ready")
+                  Task.unit
+                } else {
+                  logger.info(s"[$name] ES not ready yet, health status is $healthStatus")
+                  Task.raiseError(ClusterNotReady)
+                }
+              case Mode.Accessible =>
+                logger.info(s"[$name] ES is ready")
+                Task.unit
+            }
+          case otherStatus =>
+            logger.info(s"[$name] ES not ready yet, received HTTP $otherStatus")
+            Task.raiseError(ClusterNotReady)
+        }
+        response.close()
```

Comment:
shouldn't we introduce cats Resource somewhere here?

Reply:
done, response wrapped in Resource